File: src/main/java/org/plos/namedentity/api/IndividualComposite.java
/*
 * Copyright (c) 2006-2014 by Public Library of Science
 * http://plos.org
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.plos.namedentity.api;

import org.plos.namedentity.api.entity.*;
import org.plos.namedentity.validate.Validatable;

import javax.xml.bind.annotation.XmlElement;
import javax.xml.bind.annotation.XmlRootElement;
import javax.xml.bind.annotation.XmlTransient;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import static org.plos.namedentity.api.NedException.ErrorType.NoAuthEntity;
import static org.plos.namedentity.api.NedException.ErrorType.NoEmailEntities;
import static org.plos.namedentity.api.NedException.ErrorType.NoProfileEntities;

@XmlRootElement
public class IndividualComposite extends Composite implements Validatable {

  @XmlElement(name = "credentials")
  private List<Auth>              auth;

  private List<Individualprofile> individualprofiles;
  private List<Group>             groups;
  private List<Address>           addresses;
  private List<Email>             emails;
  private List<Phonenumber>       phonenumbers;
  private List<Uniqueidentifier>  uniqueidentifiers;
  private List<Degree>            degrees;
  private List<Url>               urls;
  private List<Relationship>      relationships;

  public static String typeName = "Individual";

  public String getTypeName() {
    return typeName;
  }

  @XmlTransient
  public Map<Class, List<? extends Entity>> getAsMap() {

    // use a linked map to preserve insertion order. this will ensure
    // that email inserts before auth record during composite creation
    // (auth record has a foreign key to email record)

    Map<Class, List<? extends Entity>> map = new LinkedHashMap<>();

    map.put(Individualprofile.class, individualprofiles);
    map.put(Group.class, groups);
    map.put(Address.class, addresses);
    map.put(Email.class, emails);
    map.put(Phonenumber.class, phonenumbers);
    map.put(Uniqueidentifier.class, uniqueidentifiers);
    map.put(Degree.class, degrees);
    map.put(Url.class, urls);
    map.put(Auth.class, auth);
    map.put(Relationship.class, relationships);

    return map;
  }

  @SuppressWarnings("unchecked")
  public void setFromMap(Map<Class, List<? extends Entity>> map) {
    individualprofiles = (List<Individualprofile>) map.get(Individualprofile.class);
    groups             = (List<Group>) map.get(Group.class);
    addresses          = (List<Address>) map.get(Address.class);
    emails             = (List<Email>) map.get(Email.class);
    phonenumbers       = (List<Phonenumber>) map.get(Phonenumber.class);
    uniqueidentifiers  = (List<Uniqueidentifier>) map.get(Uniqueidentifier.class);
    degrees            = (List<Degree>) map.get(Degree.class);
    urls               = (List<Url>) map.get(Url.class);
    auth               = (List<Auth>) map.get(Auth.class);
    relationships      = (List<Relationship>) map.get(Relationship.class);
  }

  @Override
  public void validate() {

    Map<Class, List<? extends Entity>> compositeMap = getAsMap();

    for (List<? extends Entity> entities : compositeMap.values()) {

      if (entities != null) {
        for (Entity entity : entities)
          entity.validate();
      }
    }

    if (individualprofiles == null || individualprofiles.size() == 0)
      throw new NedException(NoProfileEntities, "Profile entities can not be empty");

    if (emails == null || emails.size() == 0)
      throw new NedException(NoEmailEntities, "Email entities can not be empty");

    if (auth == null || auth.size() == 0)
      throw new NedException(NoAuthEntity, "User credentials can not be empty");
  }

  public List<Auth> getAuth() {
    return auth;
  }

  public void setAuth(List<Auth> auth) {
    this.auth = auth;
  }

  public List<Individualprofile> getIndividualprofiles() {
    return individualprofiles;
  }

  public void setIndividualprofiles(List<Individualprofile> individualprofiles) {
    this.individualprofiles = individualprofiles;
  }

  public List<Url> getUrls() {
    return urls;
  }

  public void setUrls(List<Url> urls) {
    this.urls = urls;
  }

  public List<Degree> getDegrees() {
    return degrees;
  }

  public void setDegrees(List<Degree> degrees) {
    this.degrees = degrees;
  }

  public List<Group> getGroups() {
    return groups;
  }

  public void setGroups(List<Group> groups) {
    this.groups = groups;
  }

  public List<Address> getAddresses() {
    return addresses;
  }

  public void setAddresses(List<Address> addresses) {
    this.addresses = addresses;
  }

  public List<Email> getEmails() {
    return emails;
  }

  public void setEmails(List<Email> emails) {
    this.emails = emails;
  }

  public List<Phonenumber> getPhonenumbers() {
    return phonenumbers;
  }

  public void setPhonenumbers(List<Phonenumber> phonenumbers) {
    this.phonenumbers = phonenumbers;
  }

  public List<Uniqueidentifier> getUniqueidentifiers() {
    return uniqueidentifiers;
  }

  public void setUniqueidentifiers(List<Uniqueidentifier> uniqueidentifiers) {
    this.uniqueidentifiers = uniqueidentifiers;
  }

  public List<Relationship> getRelationships() {
    return relationships;
  }

  public void setRelationships(List<Relationship> relationships) {
    this.relationships = relationships;
  }

  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();

    Map<Class, List<? extends Entity>> compositeMap = getAsMap();
    for (List<? extends Entity> entities : compositeMap.values()) {
      if (entities != null) {
        for (Entity entity : entities)
          sb.append(entity).append("\n");
      }
    }
    return sb.toString();
  }
}

File: client/ruby/lib/swagger_client/models/individual_composite.rb
module SwaggerClient
  # 
  class IndividualComposite < BaseObject
    attr_accessor :individualprofiles, :groups, :addresses, :emails, :phonenumbers, :uniqueidentifiers, :degrees, :urls, :relationships, :type_name, :credentials, :as_map
    # attribute mapping from ruby-style variable name to JSON key
    def self.attribute_map
      {
        
        # 
        :'individualprofiles' => :'individualprofiles',
        
        # 
        :'groups' => :'groups',
        
        # 
        :'addresses' => :'addresses',
        
        # 
        :'emails' => :'emails',
        
        # 
        :'phonenumbers' => :'phonenumbers',
        
        # 
        :'uniqueidentifiers' => :'uniqueidentifiers',
        
        # 
        :'degrees' => :'degrees',
        
        # 
        :'urls' => :'urls',
        
        # 
        :'relationships' => :'relationships',
        
        # 
        :'type_name' => :'typeName',
        
        # 
        :'credentials' => :'credentials',
        
        # 
        :'as_map' => :'asMap'
        
      }
    end

    # attribute type
    def self.swagger_types
      {
        :'individualprofiles' => :'Array<Individualprofile>',
        :'groups' => :'Array<Group>',
        :'addresses' => :'Array<Address>',
        :'emails' => :'Array<Email>',
        :'phonenumbers' => :'Array<Phonenumber>',
        :'uniqueidentifiers' => :'Array<Uniqueidentifier>',
        :'degrees' => :'Array<Degree>',
        :'urls' => :'Array<Url>',
        :'relationships' => :'Array<Relationship>',
        :'type_name' => :'String',
        :'credentials' => :'Array<Auth>',
        :'as_map' => :'Map[Class,List[? extends org.plos.namedentity.api.entity.Entity]]'
        
      }
    end

    def initialize(attributes = {})
      return if !attributes.is_a?(Hash) || attributes.empty?

      # convert string to symbol for hash key
      attributes = attributes.inject({}){|memo,(k,v)| memo[k.to_sym] = v; memo}

      
      if attributes[:'individualprofiles']
        if (value = attributes[:'individualprofiles']).is_a?(Array)
          self.individualprofiles = value
        end
      end
      
      if attributes[:'groups']
        if (value = attributes[:'groups']).is_a?(Array)
          self.groups = value
        end
      end
      
      if attributes[:'addresses']
        if (value = attributes[:'addresses']).is_a?(Array)
          self.addresses = value
        end
      end
      
      if attributes[:'emails']
        if (value = attributes[:'emails']).is_a?(Array)
          self.emails = value
        end
      end
      
      if attributes[:'phonenumbers']
        if (value = attributes[:'phonenumbers']).is_a?(Array)
          self.phonenumbers = value
        end
      end
      
      if attributes[:'uniqueidentifiers']
        if (value = attributes[:'uniqueidentifiers']).is_a?(Array)
          self.uniqueidentifiers = value
        end
      end
      
      if attributes[:'degrees']
        if (value = attributes[:'degrees']).is_a?(Array)
          self.degrees = value
        end
      end
      
      if attributes[:'urls']
        if (value = attributes[:'urls']).is_a?(Array)
          self.urls = value
        end
      end
      
      if attributes[:'relationships']
        if (value = attributes[:'relationships']).is_a?(Array)
          self.relationships = value
        end
      end
      
      if attributes[:'typeName']
        self.type_name = attributes[:'typeName']
      end
      
      if attributes[:'credentials']
        if (value = attributes[:'credentials']).is_a?(Array)
          self.credentials = value
        end
      end
      
      if attributes[:'asMap']
        self.as_map = attributes[:'asMap']
      end
      
    end

  end
end
